### ISSUE TITLE
Computing Top-K accuracy on validation data is unproportionately slow

### ISSUE THREAD

**Issue by msvensson222** [tensorflow/recommenders#388]
I use TFRS in an e-commerce retail setting with a lot of purchase history data as well as click-stream data, using a multitask recommender model. I have created a model that works fine to train, evaluate and serve, with one unfortunate issue: validation during training takes 5x the time of the training, on a fraction of the data. To exemplify, I have 10M rows of interactions, that I split into 70/20/10 split for training/validation/test. There are 1M unique users and 100k unique items.

I train using 4 GPU's, and one epoch takes ~4 minutes to run through the 7M training rows, but another ~20 minutes to get Top-K accuracy on the validation data (2M rows). During training, I make sure to specify `compute_metrics=not training`.

My retrieval task is set up as follows, where I use as large batch-size as my hardware allows to speed it up as much as possible.
```python
self.retrieval_task: tf.keras.layers.Layer = tfrs.tasks.Retrieval(
    loss=...,
    metrics=tfrs.metrics.FactorizedTopK(candidates=items_ds.batch(8192).map(self.item_model))
)
```

And I train as follows:
```python
model.fit(train.batch(BATCH_SIZE),
          epochs=15,
          validation_data=val.batch(BATCH_SIZE)
)
```
If I run 15 epochs I spend 1 hour training and 5 hours validating, on one seventh of the data.
Is this immense time difference expected? Anything I can do to improve the performance during validation?

**Comment by patrickorlando** [comment]
@hkristof03,

The easiest way to evaluate metrics only for each validation step is to pass `compute_metrics = not training` in your call to the retrieval loss.

When the metrics are computed, the candidate embeddings are re-calculated for every validation batch. The code snippet @maciejkula posted saves computation by calculating the candidate embeddings only once per validation epoch. 

If you wish to apply this optimisation you would indeed need to periodically stop training, recompile and call evaluate.

**Comment by hkristof03** [comment]
This optimization works quite well, but it causes memory explosion, because the candidates are recomputed for each validation and for that a new computation graph is created each time ([link](https://stackoverflow.com/questions/67116476/memory-leak-for-custom-tensorflow-training-using-tf-function)). We followed this tutorial to write our custom training & test loops, they are decorated with @tf.function. Do you have any suggestion? tf.keras.backend.clear_session() & gc.collect() do not solve the problem, they only partially release some memory, as we checked with tf.config.experimental.get_memory_info('GPU:0'). I did not find any options on the web to manually clear the tensorflow function cache.

**Comment by patrickorlando** [comment]
Hi @hkristof03,

I'm certainly not an expert in this area and, you may have already implemented it in this way, but I'll share my thoughts.
It might help if you could post a code sample of the `val_step` and  where you are computing the new candidates before each validation epoch.

Since you only have 50K items, you can probably get away with the BruteForce Index.

My approach would be to:
1. Create or load your candidates in a `tf.data.Dataset`.
1. Create a BruteForce layer and index it with `index_from_dataset(candidate_ds.batch(512).map(candidate_model))`. These vectors will initially be random.
2. Define the `train_step` and `test_step` outside of your training loop. The `val_step` will use the index created above. Decorated these with tf.function.
3. Before each validation epoch, re-index the retrieval layer (don't create a new layer each time), use the same candidate_ds. Ensure this function is not wrapped in a `tf.function` decorator.

_I think_ that because, the candidates are stored as `tf.Variable`s, re-indexing will update the state of those variables, rather than creating new nodes in the graph. 
I'm not exactly sure how the ScaNN layer will behave, so I'd try the BruteForce first.

Let me know if this helps.

**Comment by patrickorlando** [comment]
@houghtonweihu You need to use the index_from_dataset method to index from a dataset.
```
brute_force = tfrs.layers.factorized_top_k.BruteForce(model.user_model)
brute_force.index_from_dataset(
    movies.batch(128).map(lambda title: (title, model.movie_model(title)))
)
```

Make sure you are referencing the tutorial that matches the tfrs version you are using. [Current Retrieval Tutorial](https://www.tensorflow.org/recommenders/examples/basic_retrieval#making_predictions)

I also find reading the function docstring to be helpful when encountering errors like this.

**Comment by mayanksingh09** [comment]
A lot of the other comments mention this but writing it here if this helps. If you don't want to be 100% exact about the accuracy measure you can use ScaNN in this manner:
```
self.scann = tfrs.layers.factorized_top_k.ScaNN()  # can change num_leaves, num_leaves_to_search to change time it takes
self.scann.index_from_dataset(item_ds.batch(8192).map(self.item_model))  # This will take some time
self.retrieval_task: tf.keras.layers.Layer = tfrs.tasks.Retrieval(
    loss=...,
    metrics=tfrs.metrics.FactorizedTopK(candidates=self.scann)
)
```

**Comment by AllardJM** [comment]
`compute_metrics = not training`

@patrickorlando I am looking for exactly this but it doesnt seem to be available. Is there a way to only compute the metrics on the validation set being passed into .fit()? The use case is for built in early stopping callback.
